# Data Preparation for Fraud Detection

This notebook is dedicated to preparing the dataset for a fraud detection model. The focus is on cleaning, transforming, and inital exploring the data to ensure it's well-suited for effective model training.

In [1]:
# import required libraries
import polars as pl

# adjust display setting polars
pl.Config.set_tbl_cols(-1)

polars.config.Config

## Initial Exploration
Begin with an initial exploratory to understand the dataset's structure, identify any immediate issues, and plan for necessary data cleaning steps.

In [2]:
# load the historical transactions
transactions_df = pl.read_csv("../raw-datasets/credit_card_transactions-ibm_v2.csv")

# display it
transactions_df

shape: (24_386_900, 15)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┬──────┬────────┬───────┐
│ Use ┆ Car ┆ Yea ┆ Mon ┆ Day ┆ Tim ┆ Amo ┆ Use ┆ Mer ┆ Mer ┆ Mer ┆ Zip    ┆ MCC  ┆ Errors ┆ Is    │
│ r   ┆ d   ┆ r   ┆ th  ┆ --- ┆ e   ┆ unt ┆ Chi ┆ cha ┆ cha ┆ cha ┆ ---    ┆ ---  ┆ ?      ┆ Fraud │
│ --- ┆ --- ┆ --- ┆ --- ┆ i64 ┆ --- ┆ --- ┆ p   ┆ nt  ┆ nt  ┆ nt  ┆ f64    ┆ i64  ┆ ---    ┆ ?     │
│ i64 ┆ i64 ┆ i64 ┆ i64 ┆     ┆ str ┆ str ┆ --- ┆ Nam ┆ Cit ┆ Sta ┆        ┆      ┆ str    ┆ ---   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ str ┆ e   ┆ y   ┆ te  ┆        ┆      ┆        ┆ str   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- ┆ --- ┆ --- ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ i64 ┆ str ┆ str ┆        ┆      ┆        ┆       │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╪══════╪════════╪═══════╡
│ 0   ┆ 0   ┆ 200 ┆ 9   ┆ 1   ┆ 06: ┆ $13 ┆ Swi ┆ 352 ┆ La  ┆ CA  ┆ 91750. ┆ 5300 ┆ null   ┆ No    │
│     ┆     ┆ 2   ┆     ┆     ┆ 21  ┆ 4.0 ┆ pe  ┆ 721 ┆ Ver ┆     ┆ 0      ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆ 9   ┆ Tra ┆ 324 ┆ ne  ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ nsa ┆ 612 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cti ┆ 787 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ on  ┆ 695 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆        ┆      ┆        ┆       │
│ 0   ┆ 0   ┆ 200 ┆ 9   ┆ 1   ┆ 06: ┆ $38 ┆ Swi ┆ -72 ┆ Mon ┆ CA  ┆ 91754. ┆ 5411 ┆ null   ┆ No    │
│     ┆     ┆ 2   ┆     ┆     ┆ 42  ┆ .48 ┆ pe  ┆ 761 ┆ ter ┆     ┆ 0      ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ Tra ┆ 209 ┆ ey  ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ nsa ┆ 213 ┆ Par ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cti ┆ 991 ┆ k   ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ on  ┆ 604 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆        ┆      ┆        ┆       │
│ 0   ┆ 0   ┆ 200 ┆ 9   ┆ 2   ┆ 06: ┆ $12 ┆ Swi ┆ -72 ┆ Mon ┆ CA  ┆ 91754. ┆ 5411 ┆ null   ┆ No    │
│     ┆     ┆ 2   ┆     ┆     ┆ 22  ┆ 0.3 ┆ pe  ┆ 761 ┆ ter ┆     ┆ 0      ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ Tra ┆ 209 ┆ ey  ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ nsa ┆ 213 ┆ Par ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cti ┆ 991 ┆ k   ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ on  ┆ 604 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆        ┆      ┆        ┆       │
│ 0   ┆ 0   ┆ 200 ┆ 9   ┆ 2   ┆ 17: ┆ $12 ┆ Swi ┆ 341 ┆ Mon ┆ CA  ┆ 91754. ┆ 5651 ┆ null   ┆ No    │
│     ┆     ┆ 2   ┆     ┆     ┆ 45  ┆ 8.9 ┆ pe  ┆ 452 ┆ ter ┆     ┆ 0      ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆ 5   ┆ Tra ┆ 745 ┆ ey  ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ nsa ┆ 957 ┆ Par ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cti ┆ 910 ┆ k   ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ on  ┆ 677 ┆     ┆     ┆        ┆      ┆        ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 0   ┆     ┆     ┆        ┆      ┆        ┆       │
│ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …   ┆ …      ┆ …    ┆ …      ┆ …     │
│ 199 ┆ 1   ┆ 202 ┆ 2   ┆ 27  ┆ 22: ┆ $54 ┆ Chi ┆ -51 ┆ Mer ┆ NH  ┆ 3054.0 ┆ 5541 ┆ null   ┆ No    │
│ 9   ┆     ┆ 0   ┆     ┆     ┆ 24  ┆

Overview: 
With approximately 24.4 million transactions, this dataset offers a comprehensive set of data points. The upcoming steps involve thorough data cleaning and transformation to shape the data for optimal use in machine learning model training.

We'll conduct an initial exploration to get a sense of the dataset's quality, identify any missing values, and understand the overall data distribution.

In [3]:
# calculate of missing values in each column
missing_values = transactions_df.select([pl.col(column).is_null().sum().alias(column) for column in transactions_df.columns])
print("Percentage of Missing Values:")
missing_values / len(transactions_df) * 100

Percentage of Missing Values:


shape: (1, 15)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┬───────┬─────┬───────┬───────┐
│ Use ┆ Car ┆ Yea ┆ Mon ┆ Day ┆ Tim ┆ Amo ┆ Use ┆ Mer ┆ Mer ┆ Mercha ┆ Zip   ┆ MCC ┆ Error ┆ Is    │
│ r   ┆ d   ┆ r   ┆ th  ┆ --- ┆ e   ┆ unt ┆ Chi ┆ cha ┆ cha ┆ nt     ┆ ---   ┆ --- ┆ s?    ┆ Fraud │
│ --- ┆ --- ┆ --- ┆ --- ┆ f64 ┆ --- ┆ --- ┆ p   ┆ nt  ┆ nt  ┆ State  ┆ f64   ┆ f64 ┆ ---   ┆ ?     │
│ f64 ┆ f64 ┆ f64 ┆ f64 ┆     ┆ f64 ┆ f64 ┆ --- ┆ Nam ┆ Cit ┆ ---    ┆       ┆     ┆ f64   ┆ ---   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f64 ┆ e   ┆ y   ┆ f64    ┆       ┆     ┆       ┆ f64   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- ┆ --- ┆        ┆       ┆     ┆       ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f64 ┆ f64 ┆        ┆       ┆     ┆       ┆       │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╪═══════╪═════╪═══════╪═══════╡
│ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 11.156 ┆ 11.80 ┆ 0.0 ┆ 98.40 ┆ 0.0   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 896    ┆ 1972  ┆     ┆ 7215  ┆       │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴────────┴───────┴─────┴───────┴───────┘

We observed a significant number of missing values in the `Errors?` column. It seems likely that these null values indicate transactions without any errors, implying successful transactions.

In [4]:
print("`Is Fraud?` proportion: \n",
      transactions_df.group_by("Is Fraud?")
      .agg(pl.count().alias("count"))
      .with_columns([
          (pl.col("count") / len(transactions_df) * 100).alias("percentage")
          ])
      .sort("count", descending=True))

`Is Fraud?` proportion: 
 shape: (2, 3)
┌───────────┬──────────┬────────────┐
│ Is Fraud? ┆ count    ┆ percentage │
│ ---       ┆ ---      ┆ ---        │
│ str       ┆ u32      ┆ f64        │
╞═══════════╪══════════╪════════════╡
│ No        ┆ 24357143 ┆ 99.87798   │
│ Yes       ┆ 29757    ┆ 0.12202    │
└───────────┴──────────┴────────────┘


Our target column, `Is Fraud?`, shows a distribution of 99.87% (24,357,143 transactions) as non-fraudulent and only 0.13% (29,757 transactions) as fraudulent. This distribution aligns well with real-world scenarios where fraudulent transactions are relatively rare.

In [5]:
print("`Use Chip` proportion: \n",
      transactions_df.group_by("Use Chip")
      .agg(pl.count().alias("count"))
      .with_columns([
          (pl.col("count") / len(transactions_df) * 100).alias("percentage")
          ])
      .sort("count", descending=True))

`Use Chip` proportion: 
 shape: (3, 3)
┌────────────────────┬──────────┬────────────┐
│ Use Chip           ┆ count    ┆ percentage │
│ ---                ┆ ---      ┆ ---        │
│ str                ┆ u32      ┆ f64        │
╞════════════════════╪══════════╪════════════╡
│ Swipe Transaction  ┆ 15386082 ┆ 63.091586  │
│ Chip Transaction   ┆ 6287598  ┆ 25.782687  │
│ Online Transaction ┆ 2713220  ┆ 11.125727  │
└────────────────────┴──────────┴────────────┘


The `Use Chip` column represents the transaction method used by the customer. Our analysis revealed that Swipe Transactions account for 63.09%, Chip Transactions for 25.79%, and Online Transactions for 11.12%.

In [6]:
print("`Merchant State` proportion: \n",
      transactions_df.group_by("Merchant State")
      .agg(pl.count().alias("count"))
      .with_columns([
          (pl.col("count") / len(transactions_df) * 100).alias("percentage")
          ])
      .sort("count", descending=True))

`Merchant State` proportion: 
 shape: (224, 3)
┌────────────────┬─────────┬────────────┐
│ Merchant State ┆ count   ┆ percentage │
│ ---            ┆ ---     ┆ ---        │
│ str            ┆ u32     ┆ f64        │
╞════════════════╪═════════╪════════════╡
│ null           ┆ 2720821 ┆ 11.156896  │
│ CA             ┆ 2591830 ┆ 10.62796   │
│ TX             ┆ 1793298 ┆ 7.35353    │
│ FL             ┆ 1458699 ┆ 5.981486   │
│ …              ┆ …       ┆ …          │
│ Tonga          ┆ 2       ┆ 0.000008   │
│ Botswana       ┆ 1       ┆ 0.000004   │
│ Kiribati       ┆ 1       ┆ 0.000004   │
│ Paraguay       ┆ 1       ┆ 0.000004   │
└────────────────┴─────────┴────────────┘


In the `Merchant State` column, we noticed that null values comprise the largest proportion (11.15%), followed by California (10.62%) and Texas (7.3%). With approximately 224 unique values, it's clear that transactions occur not only in the USA but also in other countries, as indicated by entries like Paraguay, Togo, etc.

In [7]:
print("`Merchant City` proportion: \n",
      transactions_df.group_by("Merchant City")
      .agg(pl.count().alias("count"))
      .with_columns([
          (pl.col("count") / len(transactions_df) * 100).alias("percentage")
          ])
      .sort("count", descending=True))

`Merchant City` proportion: 
 shape: (13_429, 3)
┌───────────────┬─────────┬────────────┐
│ Merchant City ┆ count   ┆ percentage │
│ ---           ┆ ---     ┆ ---        │
│ str           ┆ u32     ┆ f64        │
╞═══════════════╪═════════╪════════════╡
│ ONLINE        ┆ 2720821 ┆ 11.156896  │
│ Houston       ┆ 246036  ┆ 1.008886   │
│ Los Angeles   ┆ 180496  ┆ 0.740135   │
│ Miami         ┆ 178653  ┆ 0.732578   │
│ …             ┆ …       ┆ …          │
│ Boyne Falls   ┆ 1       ┆ 0.000004   │
│ Weyerhaeuser  ┆ 1       ┆ 0.000004   │
│ Poyen         ┆ 1       ┆ 0.000004   │
│ Long Bottom   ┆ 1       ┆ 0.000004   │
└───────────────┴─────────┴────────────┘


The `Merchant City` column also follows a similar pattern, with 'Online' being the most common category at 11.15%. Other frequent locations include U.S. cities such as Houston, Los Angeles, and Miami. The presence of 13,429 unique city names suggests that transactions span globally, not just within the USA.

In [8]:
print("`Errors?` proportion: \n",
      transactions_df.group_by("Errors?")
      .agg(pl.count().alias("count"))
      .with_columns([
          (pl.col("count") / len(transactions_df) * 100).alias("percentage")
          ])
      .sort("count", descending=True))

`Errors?` proportion: 
 shape: (24, 3)
┌───────────────────────────────────┬──────────┬────────────┐
│ Errors?                           ┆ count    ┆ percentage │
│ ---                               ┆ ---      ┆ ---        │
│ str                               ┆ u32      ┆ f64        │
╞═══════════════════════════════════╪══════════╪════════════╡
│ null                              ┆ 23998469 ┆ 98.407215  │
│ Insufficient Balance              ┆ 242783   ┆ 0.995547   │
│ Bad PIN                           ┆ 58918    ┆ 0.241597   │
│ Technical Glitch                  ┆ 48157    ┆ 0.197471   │
│ …                                 ┆ …        ┆ …          │
│ Bad Zipcode,Insufficient Balance  ┆ 13       ┆ 0.000053   │
│ Bad Zipcode,Technical Glitch      ┆ 7        ┆ 0.000029   │
│ Bad Card Number,Bad Expiration,I… ┆ 2        ┆ 0.000008   │
│ Bad Card Number,Bad Expiration,T… ┆ 1        ┆ 0.000004   │
└───────────────────────────────────┴──────────┴────────────┘


Consistent with earlier observations, the `Errors?` column indicates that 98.40% of transactions have no recorded errors, implying successful transactions. This column also includes other error categories like insufficient balance, bad pin, etc., totaling 24 distinct categories.

In [9]:
print('Summary statistics: ')
transactions_df.describe()

Summary statistics: 


shape: (9, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┐
│ des ┆ Use ┆ Car ┆ Yea ┆ Mon ┆ Day ┆ Tim ┆ Amo ┆ Use ┆ Mer ┆ Mer ┆ Mer ┆ Zip ┆ MCC ┆ Err ┆ Is     │
│ cri ┆ r   ┆ d   ┆ r   ┆ th  ┆ --- ┆ e   ┆ unt ┆ Chi ┆ cha ┆ cha ┆ cha ┆ --- ┆ --- ┆ ors ┆ Fraud? │
│ be  ┆ --- ┆ --- ┆ --- ┆ --- ┆ f64 ┆ --- ┆ --- ┆ p   ┆ nt  ┆ nt  ┆ nt  ┆ f64 ┆ f64 ┆ ?   ┆ ---    │
│ --- ┆ f64 ┆ f64 ┆ f64 ┆ f64 ┆     ┆ str ┆ str ┆ --- ┆ Nam ┆ Cit ┆ Sta ┆     ┆     ┆ --- ┆ str    │
│ str ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ str ┆ e   ┆ y   ┆ te  ┆     ┆     ┆ str ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- ┆ --- ┆ --- ┆     ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f64 ┆ str ┆ str ┆     ┆     ┆     ┆        │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╡
│ cou ┆ 2.4 ┆ 2.4 ┆ 2.4 ┆ 2.4 ┆ 2.4 ┆ 243 ┆ 243 ┆ 243 ┆ 2.4 ┆ 243 ┆ 216 ┆ 2.1 ┆ 2.4 ┆ 388 ┆ 243869 │
│ nt  ┆ 386 ┆ 386 ┆ 386 ┆ 386 ┆ 386 ┆ 869 ┆ 869 ┆ 869 ┆ 386 ┆ 869 ┆ 660 ┆ 508 ┆ 386 ┆ 431 ┆ 00     │
│     ┆ 9e7 ┆ 9e7 ┆ 9e7 ┆ 9e7 ┆ 9e7 ┆ 00  ┆ 00  ┆ 00  ┆ 9e7 ┆ 00  ┆ 79  ┆ 765 ┆ 9e7 ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ e7  ┆     ┆     ┆        │
│ nul ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0   ┆ 0   ┆ 0   ┆ 0.0 ┆ 0   ┆ 272 ┆ 2.8 ┆ 0.0 ┆ 239 ┆ 0      │
│ l_c ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 082 ┆ 781 ┆     ┆ 984 ┆        │
│ oun ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 1   ┆ 35e ┆     ┆ 69  ┆        │
│ t   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 6   ┆     ┆     ┆        │
│ mea ┆ 100 ┆ 1.3 ┆ 201 ┆ 6.5 ┆ 15. ┆ nul ┆ nul ┆ nul ┆ -4. ┆ nul ┆ nul ┆ 509 ┆ 556 ┆ nul ┆ null   │
│ n   ┆ 1.0 ┆ 513 ┆ 1.9 ┆ 250 ┆ 718 ┆ l   ┆ l   ┆ l   ┆ 769 ┆ l   ┆ l   ┆ 56. ┆ 1.1 ┆ l   ┆        │
│     ┆ 193 ┆ 66  ┆ 551 ┆ 64  ┆ 123 ┆     ┆     ┆     ┆ 2e1 ┆     ┆     ┆ 442 ┆ 712 ┆     ┆        │
│     ┆ 35  ┆     ┆ 7   ┆     ┆     ┆     ┆     ┆     ┆ 7   ┆     ┆     ┆ 115 ┆ 53  ┆     ┆        │
│ std ┆ 569 ┆ 1.4 ┆ 5.1 ┆ 3.4 ┆ 8.7 ┆ nul ┆ nul ┆ nul ┆ 4.7 ┆ nul ┆ nul ┆ 293 ┆ 879 ┆ nul ┆ null   │
│     ┆ .46 ┆ 071 ┆ 059 ┆ 723 ┆ 940 ┆ l   ┆ l   ┆ l   ┆ 589 ┆ l   ┆ l   ┆ 97. ┆ .31 ┆ l   ┆        │
│     ┆ 115 ┆ 54  ┆ 21  ┆ 55  ┆ 73  ┆     ┆     ┆     ┆ e18 ┆     ┆     ┆ 065 ┆ 543 ┆     ┆        │
│     ┆ 7   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 949 ┆ 3   ┆     ┆        │
│ min ┆ 0.0 ┆ 0.0 ┆ 199 ┆ 1.0 ┆ 1.0 ┆ 00: ┆ $-0 ┆ Chi ┆ -9. ┆ Aar ┆ AA  ┆ 501 ┆ 171 ┆ Bad ┆ No     │
│     ┆     ┆     ┆ 1.0 ┆     ┆     ┆ 00  ┆ .00 ┆ p   ┆ 222 ┆ ons ┆     ┆ .0  ┆ 1.0 ┆ CVV ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ Tra ┆ 9e1 ┆ bur ┆     ┆     ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ nsa ┆ 8   ┆ g   ┆     ┆     ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cti ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ on  ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ 25% ┆ 510 ┆ 0.0 ┆ 200 ┆ 3.0 ┆ 8.0 ┆ nul ┆ nul ┆ nul ┆ -4. ┆ nul ┆ nul ┆ 283 ┆ 530 ┆ nul ┆ null   │
│     ┆ .0  ┆     ┆ 8.0 ┆     ┆     ┆ l   ┆ l   ┆ l   ┆ 500 ┆ l   ┆ l   ┆ 74. ┆ 0.0 ┆ l   ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 5e1 ┆     ┆     ┆ 0   ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 8   ┆     ┆     ┆     ┆     ┆     ┆        │
│ 50% ┆ 100 ┆ 1.0 ┆ 201 ┆ 7.0 ┆ 16. ┆ nul ┆ nul ┆ nul ┆ -7. ┆ nul ┆ nul ┆ 467 ┆ 549 ┆ nul ┆ null   │
│     ┆ 6.0 ┆     ┆ 3.0 ┆     ┆ 0   ┆ l   ┆ l   ┆ l   ┆ 946 ┆ l   ┆ l   ┆ 42. ┆ 9.0 ┆ l   ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 8e1 ┆     ┆     ┆ 0   ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 7   ┆     ┆     ┆     ┆     ┆     ┆        │
│ 75% ┆ 147 ┆ 2.0 ┆ 201 ┆ 10. ┆ 23. ┆ nul ┆ nu

From the summary statistics presented above, it's challenging to discern clear insights due to the raw and unclean state of the data. Post-cleanup, we will conduct another round of summary statistics to better understand the dataset.

### Data Cleaning and Features Transformation

In [10]:
# clone the dataframe
transactions_prepared = transactions_df.clone()

We are set to perform initial data cleaning tasks. These include removing unnecessary characters, typecasting columns for consistency, and creating new features to enrich our dataset.

In [11]:
# type casting, remove unnecesarry characters, create new datetime column
transactions_prepared = transactions_prepared.with_columns([
    pl.col("Amount").str.strip_chars("$").cast(pl.Float64).abs(), # remove dollar sign, cast to float, turn into absolute incase there's minus trx which is not make sense
    pl.col("Merchant Name").cast(pl.Utf8), # cast to string
    pl.col("MCC").cast(pl.Utf8).cast(pl.Categorical), # cast to categorical
    pl.col("Use Chip").cast(pl.Categorical), 
    pl.col("Errors?").cast(pl.Categorical), 

    # create 'timestamp_transaction' column
    (pl.col("Year").cast(pl.Utf8) + "-" +
     pl.col("Month").cast(pl.Utf8).str.zfill(2) + "-" +  # add leading zero
     pl.col("Day").cast(pl.Utf8).str.zfill(2) + " " + 
     pl.col("Time")).str.strptime(pl.Datetime, "%Y-%m-%d %H:%M").alias("timestamp_transaction"),
]).drop(["Year", "Month", "Day", "Time"])  # drop unnecessary columns

# extract datetime information
# fill missing values
transactions_prepared = transactions_prepared.with_columns([
    pl.col("timestamp_transaction").dt.weekday().alias("weekday"),
    pl.col("timestamp_transaction").dt.hour().alias("hour"),
    pl.col("timestamp_transaction").dt.minute().alias("minute"),
    pl.col("Errors?").fill_null("No Error"), # fill missing values as No Error
    pl.col("Merchant City").fill_null("NA"), # fill null values as NA
    pl.col("Merchant State").fill_null("NA"),
    pl.col("Zip").fill_null("NA").str.strip_chars(".0"),
])

sys:1: CategoricalRemappingWarning: Local categoricals have different encodings, expensive re-encoding is done to perform this merge operation. Consider using a StringCache or an Enum type if the categories are known in advance


In [12]:
# rename columns
transactions_prepared = transactions_prepared.rename({
    "User":"user_id", "Card":"card_index", "Amount":"trx_amount", "Use Chip":"trx_method",
    "Merchant Name":"merch_name", "Merchant State":"merch_state", "Merchant City":"merch_city", "Zip":"zip_code",
    "MCC":"merch_category_code", "Errors?":"error_status", "Is Fraud?":"is_fraud"
})

transactions_prepared.head()

shape: (5, 15)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┬────────┬──────┬───────┐
│ use ┆ car ┆ trx ┆ trx ┆ mer ┆ mer ┆ mer ┆ zip ┆ mer ┆ err ┆ is_ ┆ timest ┆ weekda ┆ hour ┆ minut │
│ r_i ┆ d_i ┆ _am ┆ _me ┆ ch_ ┆ ch_ ┆ ch_ ┆ _co ┆ ch_ ┆ or_ ┆ fra ┆ amp_tr ┆ y      ┆ ---  ┆ e     │
│ d   ┆ nde ┆ oun ┆ tho ┆ nam ┆ cit ┆ sta ┆ de  ┆ cat ┆ sta ┆ ud  ┆ ansact ┆ ---    ┆ i8   ┆ ---   │
│ --- ┆ x   ┆ t   ┆ d   ┆ e   ┆ y   ┆ te  ┆ --- ┆ ego ┆ tus ┆ --- ┆ ion    ┆ i8     ┆      ┆ i8    │
│ i64 ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- ┆ str ┆ ry_ ┆ --- ┆ str ┆ ---    ┆        ┆      ┆       │
│     ┆ i64 ┆ f64 ┆ cat ┆ str ┆ str ┆ str ┆     ┆ cod ┆ cat ┆     ┆ dateti ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ e   ┆     ┆     ┆ me[μs] ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ cat ┆     ┆     ┆        ┆        ┆      ┆       │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╪════════╪══════╪═══════╡
│ 0   ┆ 0   ┆ 134 ┆ Swi ┆ 352 ┆ La  ┆ CA  ┆ 917 ┆ 530 ┆ No  ┆ No  ┆ 2002-0 ┆ 7      ┆ 6    ┆ 21    │
│     ┆     ┆ .09 ┆ pe  ┆ 721 ┆ Ver ┆     ┆ 5   ┆ 0   ┆ Err ┆     ┆ 9-01   ┆        ┆      ┆       │
│     ┆     ┆     ┆ Tra ┆ 324 ┆ ne  ┆     ┆     ┆     ┆ or  ┆     ┆ 06:21: ┆        ┆      ┆       │
│     ┆     ┆     ┆ nsa ┆ 612 ┆     ┆     ┆     ┆     ┆     ┆     ┆ 00     ┆        ┆      ┆       │
│     ┆     ┆     ┆ cti ┆ 787 ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆ on  ┆ 695 ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│ 0   ┆ 0   ┆ 38. ┆ Swi ┆ -72 ┆ Mon ┆ CA  ┆ 917 ┆ 541 ┆ No  ┆ No  ┆ 2002-0 ┆ 7      ┆ 6    ┆ 42    │
│     ┆     ┆ 48  ┆ pe  ┆ 761 ┆ ter ┆     ┆ 54  ┆ 1   ┆ Err ┆     ┆ 9-01   ┆        ┆      ┆       │
│     ┆     ┆     ┆ Tra ┆ 209 ┆ ey  ┆     ┆     ┆     ┆ or  ┆     ┆ 06:42: ┆        ┆      ┆       │
│     ┆     ┆     ┆ nsa ┆ 213 ┆ Par ┆     ┆     ┆     ┆     ┆     ┆ 00     ┆        ┆      ┆       │
│     ┆     ┆     ┆ cti ┆ 991 ┆ k   ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆ on  ┆ 604 ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│ 0   ┆ 0   ┆ 120 ┆ Swi ┆ -72 ┆ Mon ┆ CA  ┆ 917 ┆ 541 ┆ No  ┆ No  ┆ 2002-0 ┆ 1      ┆ 6    ┆ 22    │
│     ┆     ┆ .34 ┆ pe  ┆ 761 ┆ ter ┆     ┆ 54  ┆ 1   ┆ Err ┆     ┆ 9-02   ┆        ┆      ┆       │
│     ┆     ┆     ┆ Tra ┆ 209 ┆ ey  ┆     ┆     ┆     ┆ or  ┆     ┆ 06:22: ┆        ┆      ┆       │
│     ┆     ┆     ┆ nsa ┆ 213 ┆ Par ┆     ┆     ┆     ┆     ┆     ┆ 00     ┆        ┆      ┆       │
│     ┆     ┆     ┆ cti ┆ 991 ┆ k   ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆ on  ┆ 604 ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆ 3   ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│ 0   ┆ 0   ┆ 128 ┆ Swi ┆ 341 ┆ Mon ┆ CA  ┆ 917 ┆ 565 ┆ No  ┆ No  ┆ 2002-0 ┆ 1      ┆ 17   ┆ 45    │
│     ┆     ┆ .95 ┆ pe  ┆ 452 ┆ ter ┆     ┆ 54  ┆ 1   ┆ Err ┆     ┆ 9-02   ┆        ┆      ┆       │
│     ┆     ┆     ┆ Tra ┆ 745 ┆ ey  ┆     ┆     ┆     ┆ or  ┆     ┆ 17:45: ┆        ┆      ┆       │
│     ┆     ┆     ┆ nsa ┆ 957 ┆ Par ┆     ┆     ┆     ┆     ┆     ┆ 00     ┆        ┆      ┆       │
│     ┆     ┆     ┆ cti ┆ 910 ┆ k   ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆ on  ┆ 677 ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│     ┆     ┆     ┆     ┆ 0   ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆        ┆      ┆       │
│ 0   ┆ 0   ┆ 104 ┆ Swi ┆ 581 ┆ La  ┆ CA  ┆ 91

Columns are renamed for clarity and consistency, which aids in making the dataset more understandable and easier to work with during further analysis and modeling.

In [13]:
print('Summary statistics: ')
transactions_prepared.describe()

Summary statistics: 


shape: (9, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┐
│ des ┆ use ┆ car ┆ trx ┆ trx ┆ mer ┆ mer ┆ mer ┆ zip ┆ mer ┆ err ┆ is_ ┆ tim ┆ wee ┆ hou ┆ minute │
│ cri ┆ r_i ┆ d_i ┆ _am ┆ _me ┆ ch_ ┆ ch_ ┆ ch_ ┆ _co ┆ ch_ ┆ or_ ┆ fra ┆ est ┆ kda ┆ r   ┆ ---    │
│ be  ┆ d   ┆ nde ┆ oun ┆ tho ┆ nam ┆ cit ┆ sta ┆ de  ┆ cat ┆ sta ┆ ud  ┆ amp ┆ y   ┆ --- ┆ f64    │
│ --- ┆ --- ┆ x   ┆ t   ┆ d   ┆ e   ┆ y   ┆ te  ┆ --- ┆ ego ┆ tus ┆ --- ┆ _tr ┆ --- ┆ f64 ┆        │
│ str ┆ f64 ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- ┆ str ┆ ry_ ┆ --- ┆ str ┆ ans ┆ f64 ┆     ┆        │
│     ┆     ┆ f64 ┆ f64 ┆ str ┆ str ┆ str ┆ str ┆     ┆ cod ┆ str ┆     ┆ act ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ e   ┆     ┆     ┆ ion ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- ┆     ┆     ┆ --- ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ str ┆     ┆     ┆ str ┆     ┆     ┆        │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╡
│ cou ┆ 2.4 ┆ 2.4 ┆ 2.4 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 243 ┆ 2.4 ┆ 2.4 ┆ 2.4386 │
│ nt  ┆ 386 ┆ 386 ┆ 386 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 869 ┆ 386 ┆ 386 ┆ 9e7    │
│     ┆ 9e7 ┆ 9e7 ┆ 9e7 ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 00  ┆ 9e7 ┆ 9e7 ┆        │
│ nul ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0.0 ┆ 0.0 ┆ 0.0    │
│ l_c ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ oun ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ t   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ mea ┆ 100 ┆ 1.3 ┆ 54. ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ 4.0 ┆ 12. ┆ 29.585 │
│ n   ┆ 1.0 ┆ 513 ┆ 051 ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ 034 ┆ 414 ┆ 951    │
│     ┆ 193 ┆ 66  ┆ 265 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 77  ┆ 2   ┆        │
│     ┆ 35  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ std ┆ 569 ┆ 1.4 ┆ 75. ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ 1.9 ┆ 5.0 ┆ 17.318 │
│     ┆ .46 ┆ 071 ┆ 564 ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ 997 ┆ 658 ┆ 349    │
│     ┆ 115 ┆ 54  ┆ 936 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 03  ┆ 82  ┆        │
│     ┆ 7   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ min ┆ 0.0 ┆ 0.0 ┆ 0.0 ┆ nul ┆ -10 ┆ Aar ┆ AA  ┆ 100 ┆ nul ┆ nul ┆ No  ┆ 199 ┆ 1.0 ┆ 0.0 ┆ 0.0    │
│     ┆     ┆     ┆     ┆ l   ┆ 000 ┆ ons ┆     ┆ 01  ┆ l   ┆ l   ┆     ┆ 1-0 ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ 809 ┆ bur ┆     ┆     ┆     ┆     ┆     ┆ 1-0 ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ 090 ┆ g   ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ 584 ┆     ┆     ┆     ┆     ┆     ┆     ┆ 07: ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ 899 ┆     ┆     ┆     ┆     ┆     ┆     ┆ 10: ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ 71  ┆     ┆     ┆     ┆     ┆     ┆     ┆ 00  ┆     ┆     ┆        │
│ 25% ┆ 510 ┆ 0.0 ┆ 12. ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ 2.0 ┆ 9.0 ┆ 15.0   │
│     ┆ .0  ┆     ┆ 22  ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆     ┆     ┆        │
│ 50% ┆ 100 ┆ 1.0 ┆ 36. ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ 4.0 ┆ 12. ┆ 30.0   │
│     ┆ 6.0 ┆     ┆ 17  ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆     ┆ 0   ┆        │
│ 75% ┆ 147 ┆ 2.0 ┆ 72. ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ nul ┆ 6.0 ┆ 16. ┆ 45.0   │
│     ┆ 7.0 ┆     ┆ 0   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆ l   ┆     ┆ 0   ┆        │
│ max ┆ 199 ┆ 8.0 ┆ 123 ┆ nul ┆ 999 ┆ Zwo ┆ Zi

In the prepared dataframe, the summary statistics are now more insightful, particularly for the `trx_amount` and `timestamp_transaction` columns. Notably, the transaction amount varies significantly, with the smallest being $0 and the largest reaching an impressive $12,390.50. The dataset spans a substantial period, from 1991 to 2020, indicating a wide timeframe for the transactions.

We will export this prepared dataframe as a Parquet file. Opting for Parquet format offers the benefits of space efficiency and faster I/O. The saved file will be utilized for more in-depth Exploratory Data Analysis (EDA) in a separate notebook, followed by the development of our machine learning model.

In [14]:
transactions_prepared.write_parquet("../clean-datasets/transactions_prepared.parquet")